src/game/logic/object/locomotor.h

```diff
@@ -0,0 +1,178 @@
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+#pragma once
+#include "always.h"
+#include "asciistring.h"
+#include "coord.h"
+#include "overridable.h"
+#include "snapshot.h"
+
+enum LocomotorBehaviorZ
+{
+    Z_NO_Z_MOTIVE_FORCE,
+    Z_SEA_LEVEL,
+    Z_SURFACE_RELATIVE_HEIGHT,
+    Z_ABSOLUTE_HEIGHT,
+    Z_FIXED_SURFACE_RELATIVE_HEIGHT,
+    Z_FIXED_ABSOLUTE_HEIGHT,
+    Z_FIXED_RELATIVE_TO_GROUND_AND_BUILDINGS,
+    Z_RELATIVE_TO_HIGHEST_LAYER,
+};
+
+enum LocomotorAppearance
+{
+    LOCO_LEGS_TWO,
+    LOCO_WHEELS_FOUR,
+    LOCO_TREADS,
+    LOCO_HOVER,
+    LOCO_THRUST,
+    LOCO_WINGS,
+    LOCO_CLIMBER,
+    LOCO_OTHER,
+    LOCO_MOTORCYCLE,
+};
+
+enum LocomotorGroupMovementPriority
+{
+    PRIORITY_MOVES_BACK,
+    PRIORITY_MOVES_MIDDLE,
+    PRIORITY_MOVES_FRONT,
+};
+
+class LocomotorTemplate : public Overridable
+{
+    IMPLEMENT_POOL(LocomotorTemplate);
+
+public:
+    virtual ~LocomotorTemplate() override;
+
+private:
+    Utf8String m_name;
+    int m_surfaces;
+    float m_maxSpeed;
+    float m_maxSpeedDamaged;
+    float m_minSpeed;
+    float m_maxTurnRate;
+    float m_maxTurnRateDamaged;
+    float m_acceleration;
+    float m_accelerationDamaged;
+    float m_lift;
+    float m_liftDamaged;
+    float m_braking;
+    float m_minTurnSpeed;
+    float m_preferredHeight;
+    float m_preferredHeightDamping;
+    float m_circlingRadius;
+    float m_speedLimitZ;
+    float m_extra2DFriction;
+    float m_maxThrustAngle;
+    LocomotorBehaviorZ m_behaviorZ;
+    LocomotorAppearance m_appearance;
+    LocomotorGroupMovementPriority m_groupMovementPriority;
+    float m_accelPitchLimit;
+    float m_deaccelPitchLimit;
+    float m_bounceKick;
+    float m_pitchStiffness;
+    float m_rollStiffness;
+    float m_pitchDamping;
+    float m_rollDamping;
+    float m_pitchInDirectionOfZVelFactor;
+    float m_thrustRoll;
+    float m_thrustWobbleRate;
+    float m_thrustMinWobble;
+    float m_thrustMaxWobble;
+    float m_forwardVelCoef;
+    float m_lateralVelCoef;
+    float m_forwardAccelCoef;
+    float m_lateralAccelCoef;
+    float m_uniformAxialDamping;
+    float m_turnPivotOffset;
+    int m_airborneTargetingHeight;
+    float m_closeEnoughDist;
+    bool m_closeEnoughDist3D;
+    float m_slideIntoPlaceTime;
+    bool m_locomotorWorksWhenDead;
+    bool m_allowMotiveForceWhileAirborne;
+    bool m_apply2DFrictionWhenAirborne;
+    bool m_downhillOnly;
+    bool m_stickToGround;
+    bool m_canMoveBackwards;
+    bool m_hasSuspension;
+    float m_maximumWheelExtension;
+    float m_maximumWheelCompression;
+    float m_wheelTurnAngle;
+    float m_wanderWidthFactor;
+    float m_wanderLengthFactor;
+    float m_wanderAboutPointRadius;
+    float m_rudderCorrectionDegree;
+    float m_rudderCorrectionRate;
+    float m_elevatorCorrectionDegree;
+    float m_elevatorCorrectionRate;
+    friend class Locomotor;

```

Comment:
Can we move the friend to the very top? Down here by the members it is kind of obscure.

Reply:
fixed.